shared/metrics/metrics.go

```diff
@@ -0,0 +1,211 @@
+package metrics
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+// NewMetricSet returns a new MetricSet.
+func NewMetricSet(labels map[string]string) *MetricSet {
+	out := MetricSet{set: make(map[MetricType][]Sample)}
+
+	if labels != nil {
+		out.labels = labels
+	} else {
+		out.labels = make(map[string]string)
+	}
+
+	return &out
+}
+
+// AddSamples adds samples of the type metricType to the MetricSet.
+func (m *MetricSet) AddSamples(metricType MetricType, samples ...Sample) {
+	for i := 0; i < len(samples); i++ {
+		// Add global labels to samples
+		for labelName, labelValue := range m.labels {
+			// Ensure we always have a valid Labels map
+			if samples[i].Labels == nil {
+				samples[i].Labels = make(map[string]string)
+			}
+
+			// Ignore labels with empty value
```

Comment:
With the CPU info, we have the total seconds (all CPUs) and the ones for each CPU. If we didn't ignore empty values, we'd end up with `{"cpu": ""}` for the total seconds which doesn't look good.

Reply:
Makes sense, but why are you getting empty values in the input (this might be a useful comment for the future too)?

Reply:
It's this code here:
```go
	for dev, stats := range metrics.CPU {
		cpu := ""

		if dev != "cpu" {
			fmt.Sscanf(dev, "cpu%s", &cpu)
		}

		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsIOWait, Labels: map[string]string{"cpu": cpu, "mode": "iowait"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsIRQ, Labels: map[string]string{"cpu": cpu, "mode": "irq"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsIdle, Labels: map[string]string{"cpu": cpu, "mode": "idle"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsNice, Labels: map[string]string{"cpu": cpu, "mode": "nice"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsSoftIRQ, Labels: map[string]string{"cpu": cpu, "mode": "softirq"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsSteal, Labels: map[string]string{"cpu": cpu, "mode": "steal"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsSystem, Labels: map[string]string{"cpu": cpu, "mode": "system"}})
		set.AddSamples(CPUSecondsTotal, Sample{Value: stats.SecondsUser, Labels: map[string]string{"cpu": cpu, "mode": "user"}})
	}
```

Instead of having to check if it's a non-specific CPU and then omitting the `cpu` label, I just add an empty value, and the renderer takes care of the dropping it if it's empty.

Reply:
I see thanks. A comment for that would make sense to explain why the empty label can occur, as otherwise it seems like you could be handling invalid values for some reason.

Reply:
I went over this again, and decided to fix the CPU samples instead of handling empty label values, as those were only coming from the CPU samples.